webapp/src/components/ImageFieldEditor.tsx

```diff
@@ -197,7 +226,82 @@ export class ImageFieldEditor<U extends pxt.Asset> extends React.Component<Image
         this.galleryAssets = getAssets(true, this.asset.type);
     }
 
-    protected filterAssets(assets: pxt.Asset[], type: pxt.AssetType = this.asset?.type, isGallery = false, useTags?: boolean) {
+    protected getAvailableTags(filterAssets: pxt.Asset[]) {
+        let collectedTags: string[] = [];
+
+        if (this.galleryAssets) {
+            const specialTags = ["tile", "dialog", "background"];
+            filterAssets.forEach( (asset) => {
+                if (asset.meta.tags) {
+                    asset.meta.tags.forEach(t => {
+                        let sanitizedTag = sanitize(t);
+                        if (specialTags.indexOf(sanitizedTag) == -1) {
+                            if (collectedTags.indexOf(sanitizedTag) < 0) {
+                                collectedTags.push(sanitizedTag);
+                            }
+                        }
+                    });
+                }
+            })
+
+            return collectedTags;
+        }
+        return [];
+
+        function sanitize(tag: string) {
+            let sanitizedTag = (tag.indexOf("!") === 0 || tag.indexOf("?") === 0) && tag.length > 1 ? tag.substring(1) : tag;
+            sanitizedTag = sanitizedTag.toLowerCase();
+
+            return sanitizedTag;
+        }
+
+    }
+
+    protected tagClickHandler(tag: string) {
+        let selectedTags = this.state.gallerySelectedTags;
+        const sanitizedTag = tag.toLowerCase();
+        const index = selectedTags.indexOf(sanitizedTag);
+        if (index < 0) {
+            selectedTags.push(sanitizedTag);
+        } else {
+            selectedTags.splice(index, 1);
+        }
+        this.setState({
+            gallerySelectedTags: selectedTags
+        })
+    }
+
+    protected clearFilterTags() {
+        this.setState({
+            gallerySelectedTags: []
+        });
+    }
+
+    protected toggleFilter() {
+        this.setState({
+            filterOpen: !this.state.filterOpen
+        });
+    }
+
+    protected filterAssetsByTag(assets: pxt.Asset[]) {
+        if (this.state.gallerySelectedTags.length > 0 && this.state.filterOpen) {
+            assets = assets.filter((asset) => {
+                if (!asset.meta.tags) {
+                    return false;
+                }
+                for (let i = 0; i < asset.meta.tags.length; i++) {
+                    const t = asset.meta.tags[i];
+                    if (this.state.gallerySelectedTags.indexOf(t) >= 0) {
+                        return true;
+                    }
+                }
+                return false;
```

Comment:
```suggestion
                return !!assets.meta.tags?.find(t => this.state.gallerySelectedTags.indexOf(t) >= 0);
```

nit; that should work? I wouldn't trust it without checking real quick as I did type it out in the github comment box and not vs code or anything :)